incubator/virtualcluster/pkg/vn-agent/server/server_test.go

```diff
@@ -444,3 +455,167 @@ func TestServeLogs(t *testing.T) {
 		t.Errorf("Received wrong data: %s", result)
 	}
 }
+
+func TestServeRunInContainer(t *testing.T) {
+	fv := newServerTest()
+	defer fv.Close()
+	output := "foo bar"
+	podNamespace := "other"
+	podName := "foo"
+	expectedPodNamespace := getTenantNamespace(testcerts.TenantName, podNamespace)
+	expectedPodName := getPodName(podName, expectedPodNamespace)
+	expectedContainerName := "baz"
+	expectedCommand := "ls -a"
+	fv.kubeletServer.fakeKubelet.runFunc = func(podFullName string, uid types.UID, containerName string, cmd []string) ([]byte, error) {
+		if podFullName != expectedPodName {
+			t.Errorf("expected %s, got %s", expectedPodName, podFullName)
+		}
+		if containerName != expectedContainerName {
+			t.Errorf("expected %s, got %s", expectedContainerName, containerName)
+		}
+		if strings.Join(cmd, " ") != expectedCommand {
+			t.Errorf("expected: %s, got %v", expectedCommand, cmd)
+		}
+
+		return []byte(output), nil
+	}
+
+	tenantClient, err := newTenantClient()
+	if err != nil {
+		t.Fatalf("Got tenant client: %v", err)
+	}
+
+	resp, err := tenantClient.Post(fv.testHTTPServer.URL+"/run/"+podNamespace+"/"+podName+"/"+expectedContainerName+"?cmd=ls%20-a", "", nil)
+	if err != nil {
+		t.Fatalf("Got error POSTing: %v", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		// copying the response body did not work
+		t.Errorf("Cannot copy resp: %#v", err)
+	}
+	result := string(body)
+	if result != output {
+		t.Errorf("expected %s, got %s", output, result)
+	}
+}
+
+func TestServeRunInContainerWithUID(t *testing.T) {
+	// TODO: support UID naming translate
+}
+
+func setPodByNameFunc(fv *serverTestFramework, namespace, pod, container string) {
```

Comment:
Just curious, why do you extract this method out? the same for setGetContainerLogsFunc? 
(you did not do this for runFunc in TestServeRunInContainer)

Reply:
> Just curious, why do you extract this method out? the same for setGetContainerLogsFunc?
> (you did not do this for runFunc in TestServeRunInContainer)

there is a validation in the kubelet server, it gets the Pod spec from `GetPodByName` interface and check whether there exists the correspond container. We inject this info from `setPodByNameFunc` this method. However, there is no this validation in the `/run` api.  It's strange...